File: tasks/implementations/node/tasks/binary_search_tree.js
const fs = require('fs');
class Node {
  constructor(value) { this.value = value; this.left = null; this.right = null; }
}
function insert(root, value) {
  if (!root) return new Node(value);
  let node = root;
  while (true) {
    if (value < node.value) {
      if (!node.left) { node.left = new Node(value); return root; }
      node = node.left;
    } else if (value > node.value) {
      if (!node.right) { node.right = new Node(value); return root; }
      node = node.right;
    } else {
      return root;
    }
  }
}
function contains(root, value) {
  let node = root;
  while (node) {
    if (value < node.value) node = node.left;
    else if (value > node.value) node = node.right;
    else return true;
  }
  return false;
}
module.exports = function(size, fixturesRoot) {
  const tokens = fs.readFileSync(`${fixturesRoot}/generated/bst/${size}.txt`, 'utf8').trim().split(/\s+/).map(Number);
  const insertCount = tokens[0];
  const inserts = tokens.slice(1, 1 + insertCount);
  const queryIndex = 1 + insertCount;
  const queryCount = tokens[queryIndex];
  const queries = tokens.slice(queryIndex + 1, queryIndex + 1 + queryCount);
  let root = null;
  for (const value of inserts) root = insert(root, value);
  let total = 0;
  for (const value of queries) if (contains(root, value)) total += value;
  console.log(total);
};

File: tasks/implementations/python2/tasks/binary_search_tree.py
from tasks.compat import emit, join_path, read_text


class Node(object):
    __slots__ = ('value', 'left', 'right')

    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def _insert(root, value):
    if root is None:
        return Node(value)
    node = root
    while True:
        if value < node.value:
            if node.left is None:
                node.left = Node(value)
                return root
            node = node.left
        elif value > node.value:
            if node.right is None:
                node.right = Node(value)
                return root
            node = node.right
        else:
            return root


def _contains(root, value):
    node = root
    while node is not None:
        if value < node.value:
            node = node.left
        elif value > node.value:
            node = node.right
        else:
            return True
    return False


def run(size, fixtures_root):
    tokens = [int(x) for x in read_text(join_path(fixtures_root, 'generated', 'bst', '%s.txt' % size)).split()]
    insert_count = tokens[0]
    inserts = tokens[1:1 + insert_count]
    query_idx = 1 + insert_count
    query_count = tokens[query_idx]
    queries = tokens[query_idx + 1:query_idx + 1 + query_count]
    root = None
    for value in inserts:
        root = _insert(root, value)
    total = 0
    for value in queries:
        if _contains(root, value):
            total += value
    emit(total)
